parcels/particleset/baseparticleset.py

```diff
@@ -433,12 +447,36 @@ def execute(self, pyfunc=AdvectionRK4, endtime=None, runtime=None, dt=1.,
                                 'to a NetCDF file during the run.' % output_file.tempwritedir_base)
                 pbar = self.__create_progressbar(_starttime, endtime)
                 verbose_progress = True
+
             if dt > 0:
-                time = min(next_prelease, next_input, next_output, next_movie, next_callback, endtime)
+                next_time = min(next_prelease, next_input, next_output, next_movie, next_callback, endtime)
+            else:
+                next_time = max(next_prelease, next_input, next_output, next_movie, next_callback, endtime)
+
+            # If we don't perform interaction, only execute the normal kernel efficiently.
+            if self.interaction_kernel is None:
+                self.kernel.execute(self, endtime=next_time, dt=dt, recovery=recovery, output_file=output_file,
+                                    execute_once=execute_once)
+            # Interaction: interleave the interaction and non-interaction kernel for each time step.
+            # E.g. Inter -> Normal -> Inter -> Normal if endtime-time == 2*dt
             else:
-                time = max(next_prelease, next_input, next_output, next_movie, next_callback, endtime)
-            self.kernel.execute(self, endtime=time, dt=dt, recovery=recovery, output_file=output_file,
-                                execute_once=execute_once)
+                cur_time = time
+                while (cur_time < next_time and dt > 0) or (cur_time > next_time and dt < 0) or dt == 0:
```

Comment:
I'd still like to see that benchmarked before merge. I'll try to give it a shot this week by checking it out on GEMINI and run the perlin-field Benchmark of AOS of the main- and this branch, and see what comes out.

Reply:
@CKehl Thanks, that would be really nice.

Reply:
k, this part is roughly okay (judging from benchmarks)